src/box/txn_limbo.c

```diff
@@ -53,6 +53,9 @@ txn_limbo_create(struct txn_limbo *limbo)
 	limbo->promote_greatest_term = 0;
 	latch_create(&limbo->promote_latch);
 	limbo->confirmed_lsn = 0;
+	limbo->confirmed_entry =
+		rlist_entry(&limbo->queue, struct txn_limbo_entry, in_queue);
```

Comment:
This is not the same as cast `list<object*>*` to `object*`, because `rlist` is an intrusive list.
`rlist_entry` simply shifts the pointer to {offset of member} bytes back. `rlist` head is the same anchor as any that is inside a list element. So there's nothing stopping us from thinking that head is an anchor inside some dummy element. I never dereference it, so there's no problem.
```c
/**
 * return entry by list item
 */
#define rlist_entry(item, type, member) ({				\
	(type *)( (char *)(item) - offsetof(type,member));		\
})
```
https://github.com/tarantool/small/blob/90a365fdb65da3ffc7b233fe1777a5055fda4ce7/include/small/rlist.h#L275-L280

Reply:
A hack, but an interesting one for sure.